File: qc-core/src/main/java/com/eager/questioncloud/core/domain/point/model/ChargePointPayment.java
package com.eager.questioncloud.core.domain.point.model;

import com.eager.questioncloud.core.domain.point.enums.ChargePointPaymentStatus;
import com.eager.questioncloud.core.domain.point.enums.ChargePointType;
import com.eager.questioncloud.core.exception.CoreException;
import com.eager.questioncloud.core.exception.Error;
import com.eager.questioncloud.core.exception.InvalidPaymentException;
import java.time.LocalDateTime;
import lombok.Builder;
import lombok.Getter;

@Getter
public class ChargePointPayment {
    private final Long id;
    private final String paymentId;
    private final Long userId;
    private String receiptUrl;
    private final ChargePointType chargePointType;
    private ChargePointPaymentStatus chargePointPaymentStatus;
    private final LocalDateTime createdAt;
    private LocalDateTime paidAt;

    @Builder
    public ChargePointPayment(Long id, String paymentId, Long userId, String receiptUrl, ChargePointType chargePointType,
        ChargePointPaymentStatus chargePointPaymentStatus, LocalDateTime createdAt, LocalDateTime paidAt) {
        this.id = id;
        this.paymentId = paymentId;
        this.userId = userId;
        this.receiptUrl = receiptUrl;
        this.chargePointType = chargePointType;
        this.chargePointPaymentStatus = chargePointPaymentStatus;
        this.createdAt = createdAt;
        this.paidAt = paidAt;
    }


    public static ChargePointPayment order(String paymentId, Long userId, ChargePointType chargePointType) {
        return ChargePointPayment.builder()
            .paymentId(paymentId)
            .userId(userId)
            .chargePointType(chargePointType)
            .chargePointPaymentStatus(ChargePointPaymentStatus.ORDERED)
            .createdAt(LocalDateTime.now())
            .build();
    }

    public void approve(String receiptUrl) {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.PAID;
        this.receiptUrl = receiptUrl;
        this.paidAt = LocalDateTime.now();
    }

    public void fail() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.Fail;
    }

    public void validatePayment(int paidAmount) {
        validateStatus();
        validateAmount(paidAmount);
    }

    private void validateAmount(int paidAmount) {
        if (chargePointType.getAmount() != paidAmount) {
            throw new InvalidPaymentException();
        }
    }

    private void validateStatus() {
        if (!chargePointPaymentStatus.equals(ChargePointPaymentStatus.ORDERED)) {
            throw new CoreException(Error.ALREADY_PROCESSED_PAYMENT);
        }
    }
}

File: qc-core/src/main/java/com/eager/questioncloud/core/domain/point/model/ChargePointPayment.kt
package com.eager.questioncloud.core.domain.point.model

import com.eager.questioncloud.core.domain.point.enums.ChargePointPaymentStatus
import com.eager.questioncloud.core.domain.point.enums.ChargePointType
import com.eager.questioncloud.core.exception.CoreException
import com.eager.questioncloud.core.exception.Error
import com.eager.questioncloud.core.exception.InvalidPaymentException
import io.hypersistence.tsid.TSID
import java.time.LocalDateTime

class ChargePointPayment(
    val orderId: String,
    var paymentId: String? = null,
    val userId: Long,
    val chargePointType: ChargePointType,
    var chargePointPaymentStatus: ChargePointPaymentStatus = ChargePointPaymentStatus.ORDERED,
    val createdAt: LocalDateTime = LocalDateTime.now(),
    var requestAt: LocalDateTime? = null,
) {
    fun prepare(paymentId: String) {
        this.paymentId = paymentId
        this.chargePointPaymentStatus = ChargePointPaymentStatus.PAYMENT_REQUEST
        this.requestAt = LocalDateTime.now()
    }

    fun charge() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.CHARGED
    }

    fun cancel() {
        this.chargePointPaymentStatus = ChargePointPaymentStatus.CANCELED
    }

    fun validatePayment(paidAmount: Int) {
        validateStatus()
        validateAmount(paidAmount)
    }

    private fun validateAmount(paidAmount: Int) {
        if (chargePointType.amount != paidAmount) {
            throw InvalidPaymentException()
        }
    }

    private fun validateStatus() {
        if (chargePointPaymentStatus != ChargePointPaymentStatus.ORDERED) {
            throw CoreException(Error.ALREADY_PROCESSED_PAYMENT)
        }
    }

    companion object {
        fun createOrder(
            userId: Long,
            chargePointType: ChargePointType
        ): ChargePointPayment {
            return ChargePointPayment(
                orderId = TSID.Factory.getTsid().toString(),
                userId = userId,
                chargePointType = chargePointType
            )
        }
    }
}
